By default, EOCs train proficiencies normally

anothersimulacrum opened the pull request:
#### Summary
Infrastructure "EOC proficiency adjustments use normal training route by default"

#### Purpose of change
https://github.com/CleverRaven/Cataclysm-DDA/pull/72693 made me realize that EOCs bypass the normal proficiency training route. Make it not do that, by default.

#### Describe the solution
Add a new keyword argument to `[un]_proficiency`, `direct`, which controls if proficiency learning is done through the normal path (and so affected by focus) or by setting values directly. By default, it is done through the normal path.

When it is done through the normal path (`direct` is `false`), take the difference between the new value and the old value, and practice the proficiency for that many turns. If this would be negative, send out a debugmsg and make no adjustments.

When `direct` is `true`, work as before and adjust manually.

#### Describe alternatives you've considered
~~I'm not sure how idiomatic the debugmsg on invalid value is, or the true/false kwarg. If there's a more idiomatic way, I'd like to see it.~~
andrei suggested improvements to make it more idiomatic.

It probably makes sense for direct to only apply on `+=`, but it doesn't look like that's a feature of the system.

#### Testing
See added test, existing tests.

#### Additional context
I didn't see anywhere outside of the tests that this looked like the wrong thing to do, but I could be wrong.

anothersimulacrum committed:
By default, EOCs train proficiencies normally

Instead of bypassing the normal training methods, by default train the
difference between the old value and the new value. If an argument is
added, revert to the old behavior (and allow silly things like negative
training).

This is scruffy and probably not very idiomatic for EOC stuff.

andrei8l commented:
> I'm not sure how idiomatic the debugmsg on invalid value is, or the true/false kwarg. If there's a more idiomatic way, I'd like to see it.

<details>
<summary>
Try this: add `true` and `false` as mathematical constants and evaluate the kwarg as a double
</summary>

```diff
diff --git a/data/json/npcs/TALK_TEST.json b/data/json/npcs/TALK_TEST.json
index 75779c4b9e..c3ebc23575 100644
--- a/data/json/npcs/TALK_TEST.json
+++ b/data/json/npcs/TALK_TEST.json
@@ -1210,17 +1210,17 @@
       {
         "text": "Sets Test Proficiency learning done to 12 hours total.",
         "topic": "TALK_DONE",
-        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': 'true')", "=", "time('12h')" ] }
+        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': true)", "=", "time('12h')" ] }
       },
       {
         "text": "Sets Test Proficiency learning done to -1.",
         "topic": "TALK_DONE",
-        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': 'true')", "=", "-1" ] }
+        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': true)", "=", "-1" ] }
       },
       {
         "text": "Sets Test Proficiency learning done to 24h.",
         "topic": "TALK_DONE",
-        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': 'true')", "=", "time('24h')" ] }
+        "effect": { "math": [ "u_proficiency('prof_test', 'format': 'time_spent', 'direct': true)", "=", "time('24h')" ] }
       },
       {
         "text": "Learns Test Proficiency for 1h",
diff --git a/data/mods/TEST_DATA/EOC.json b/data/mods/TEST_DATA/EOC.json
index 1cb8566f91..fdb4692462 100644
--- a/data/mods/TEST_DATA/EOC.json
+++ b/data/mods/TEST_DATA/EOC.json
@@ -338,15 +338,15 @@
     "effect": [
       { "set_string_var": "prof_test", "target_var": { "global_val": "prof_id" } },
       { "math": [ "key_total_time_required", "=", "u_proficiency(prof_id, 'format': 'total_time_required')" ] },
-      { "math": [ "u_proficiency(prof_id, 'direct': 'true')", "=", "50" ] },
+      { "math": [ "u_proficiency(prof_id, 'direct': true)", "=", "50" ] },
       { "math": [ "key_time_spent_50", "=", "u_proficiency(prof_id)" ] },
-      { "math": [ "u_proficiency(prof_id, 'format': 'percent', 'direct': 'true')", "=", "50" ] },
+      { "math": [ "u_proficiency(prof_id, 'format': 'percent', 'direct': true)", "=", "50" ] },
       { "math": [ "key_percent_50", "=", "u_proficiency(prof_id, 'format': 'percent')" ] },
       { "math": [ "key_percent_50_turn", "=", "u_proficiency(prof_id)" ] },
-      { "math": [ "u_proficiency(prof_id, 'format': 'permille', 'direct': 'true')", "=", "50" ] },
+      { "math": [ "u_proficiency(prof_id, 'format': 'permille', 'direct': true)", "=", "50" ] },
       { "math": [ "key_permille_50", "=", "u_proficiency(prof_id, 'format': 'permille')" ] },
       { "math": [ "key_permille_50_turn", "=", "u_proficiency(prof_id)" ] },
-      { "math": [ "u_proficiency(prof_id, 'format': 'time_left', 'direct': 'true')", "=", "50" ] },
+      { "math": [ "u_proficiency(prof_id, 'format': 'time_left', 'direct': true)", "=", "50" ] },
       { "math": [ "key_time_left_50", "=", "u_proficiency(prof_id, 'format': 'time_left')" ] },
       { "math": [ "key_time_left_50_turn", "=", "u_proficiency(prof_id)" ] }
     ]
diff --git a/doc/NPCs.md b/doc/NPCs.md
index 9590151e45..b163f4f59d 100644
--- a/doc/NPCs.md
+++ b/doc/NPCs.md
@@ -1352,7 +1352,7 @@ _some functions support array arguments or kwargs, denoted with square brackets
 | moon_phase()     |  ✅  |   ❌   | N/A<br/>(global)  | Returns current phase of the Moon. <pre>MOON_NEW = 0,<br/>WAXING_CRESCENT = 1,<br/>HALF_MOON_WAXING = 2,<br/>WAXING_GIBBOUS = 3,<br/>FULL = 4,<br/>WANING_GIBBOUS = 5,<br/>HALF_MOON_WANING = 6,<br/>WANING_CRESCENT = 7 |
 | num_input(`s`/`v`,`d`/`v`)   |  ✅  |   ❌   | N/A<br/>(global)  | Prompt the player for a number.<br/>Arguments are Prompt text, Default Value:<br/>`"math": [ "u_value_to_set", "=", "num_input('Playstyle Perks Cost?', 4)" ]`|
 | pain()     |  ✅  |   ✅   | u, n  | Return or set pain<br/> Example:<br/>`{ "math": [ "n_pain()", "=", "u_pain() + 9000" ] }`|
-| proficiency(`s`/`v`)    |  ✅   |   ✅  | u, n  | Return or set proficiency<br/>Argument is proficiency ID.<br/><br/> Optional kwargs:<br/>`format`: `s` - `percent` return or set how many percent done the learning is. `permille` does likewise for permille. `time_spent` return or set total time spent. `time_left` return or set the remaining time. `total_time_required` return total time required to train a given proficiency (read only).<br/>`direct`: `s` - `false` (default) perform the adjustment by practicing the proficiency for the given amount of time. This will likely result in different values than specified. `true` perform the adjustment directly, bypassing other factors that may affect it.<br/><br/>Example:<br/>`{ "math": [ "u_proficiency('prof_intro_chemistry', 'format': 'percent')", "=", "50" ] }`|
+| proficiency(`s`/`v`)    |  ✅   |   ✅  | u, n  | Return or set proficiency<br/>Argument is proficiency ID.<br/><br/> Optional kwargs:<br/>`format`: `s` - `percent` return or set how many percent done the learning is. `permille` does likewise for permille. `time_spent` return or set total time spent. `time_left` return or set the remaining time. `total_time_required` return total time required to train a given proficiency (read only).<br/>`direct`: `true`/`false`/`d` - false (default) perform the adjustment by practicing the proficiency for the given amount of time. This will likely result in different values than specified. `true` perform the adjustment directly, bypassing other factors that may affect it.<br/><br/>Example:<br/>`{ "math": [ "u_proficiency('prof_intro_chemistry', 'format': 'percent')", "=", "50" ] }`|
 | school_level(`s`/`v`)    |  ✅   |   ❌  | u, n  | Return the highest level of spells known of that school.<br/>Argument is school ID.<br/><br/>Example:<br/>`"condition": { "math": [ "u_school_level('MAGUS')", ">=", "3"] }`|
 | school_level_adjustment(`s`/`v`)    |   ✅  |  ✅   | u, n  | Return or set temporary caster level adjustment. Only useable by EoCs that trigger on the event `opens_spellbook`. Old values will be reset to 0 before the event triggers. To avoid overwriting values from other EoCs, it is recommended to adjust the values here with `+=` or `-=` instead of setting it to an absolute value.<br/>Argument is school ID.<br/><br/>Example:<br/>`{ "math": [ "u_school_level_adjustment('MAGUS')", "+=", "3"] }`|
 | skill(`s`/`v`)    |  ✅  |   ✅   | u, n  | Return or set skill level<br/><br/>Example:<br/>`"condition": { "math": [ "u_skill('driving')", ">=", "5"] }`<br/>`"condition": { "math": [ "u_skill(someskill)", ">=", "5"] }`|
diff --git a/src/math_parser_diag.cpp b/src/math_parser_diag.cpp
index 0d57e8e498..cfc188686d 100644
--- a/src/math_parser_diag.cpp
+++ b/src/math_parser_diag.cpp
@@ -66,6 +66,11 @@ bool is_beta( char scope )
     }
 }
 
+constexpr bool is_true( double dbl )
+{
+    return dbl >= 1 || float_equals( dbl, 1 );
+}
+
 template<typename T>
 constexpr std::string_view _str_type_of()
 {
@@ -1206,7 +1211,7 @@ std::function<void( dialogue &, double )> proficiency_ass( char scope,
     double val ) {
         proficiency_id prof( prof_value.str( d ) );
         std::string const format = fmt_val.str( d );
-        std::string direct = direct_val.str( d );
+        bool const direct = is_true( direct_val.dbl( d ) );
         int to_write = 0;
         if( format == "percent" ) {
             to_write = to_turns<int>( prof->time_to_learn() * val ) / 100;
@@ -1222,14 +1227,11 @@ std::function<void( dialogue &, double )> proficiency_ass( char scope,
         }
         int before = to_turns<int>( d.actor( beta )->proficiency_practiced_time( prof ) );
         int learned = to_write - before;
-        if( direct != "true" && direct != "false" ) {
-            debugmsg( R"(Unknown direct value "%s" for proficiency, assuming "false")", direct );
-            direct = "false";
-        } else if( direct == "false" && learned < 0 ) {
+        if( !direct && learned < 0 ) {
             debugmsg( "For proficiency %s in dialogue, trying to learn negative without direct", prof.str() );
             return 0;
         }
-        if( direct == "false" ) {
+        if( !direct ) {
             d.actor( beta )->train_proficiency_for( prof, learned );
         } else {
             d.actor( beta )->set_proficiency_practiced_time( prof, to_write );
diff --git a/src/math_parser_func.h b/src/math_parser_func.h
index a205790f76..6c3a6febae 100644
--- a/src/math_parser_func.h
+++ b/src/math_parser_func.h
@@ -171,10 +171,12 @@ constexpr double e_v = 2.7182818284590452354;
 #endif
 } // namespace math_constants
 
-constexpr std::array<math_const, 3> constants{
+constexpr std::array<math_const, 5> constants{
     math_const{ "π", math_constants::pi_v },
     math_const{ "pi", math_constants::pi_v },
     math_const{ "e", math_constants::e_v },
+    math_const{ "true", 1 },
+    math_const{ "false", 0 },
 };
 
 std::vector<std::string_view> tokenize( std::string_view str, std::string_view separators,
diff --git a/tests/npc_talk_test.cpp b/tests/npc_talk_test.cpp
index f1b6e9ccc6..5c581c9e77 100644
--- a/tests/npc_talk_test.cpp
+++ b/tests/npc_talk_test.cpp
@@ -1495,7 +1495,7 @@ TEST_CASE( "npc_arithmetic", "[npc_talk]" )
                        proficiency_prof_test ) == 0 );
 
 
-    // Proficency learning without 'direct': 'true' is impacted by focus
+    // Proficency learning without 'direct': true is impacted by focus
     const auto prof_xp = [&player_character]() {
         return to_seconds<int>( player_character.get_proficiency_practiced_time( proficiency_prof_test ) );
     };
```

</details>

andrei8l committed:
More idiomatic true/false EOC argument

Add true/false as mathematical constants, and use a double for the
true/false keyword argument.
